### ISSUE TITLE
[RFC] Adding MKL-DNN Int8 functions to PyTorch/Aten/JIT backend

### ISSUE THREAD

**Issue by Jianhui-Li** [pytorch/pytorch#21120]
Migrate the Caffe2/MKL-DNN int8 operation to support Aten/JIT backend and align with Qint8 direction in Pytorch/Aten 

Motivation

With Cascadelake/VNNI, MKL-DNN int8 functions can speedup DL models can up to 4x. With this proposal, we plan to make IA's int8 function available to broader Pytorch community through integrating to the Pytorch/Aten/JIT backend following the Qint8 tensor design.  The MKL-DNN Int8 operations are already in the caffe2 backend, and this proposal is to bring the capability to Aten backend. The MKL-DNN int8 functions are tuned for Skylake and Cascadelake/VNNI.
Performance result varies according to the workload/hardware/environment setting. For vision models, we expect ~1.3x for skylake and ~2x for cascadelake.  For particular models like Resnet-50, we expect higher gains like ~1.8x and ~3x. 

Pitch

Speedup Pytorch model on IA with latest CPU instructions for deep learning (AVX512 and VNNI) 

Additional context

We plan to extend the MKLDNN opaque tensor design to support a new tensor type (Qint8) so with a new tensor type id: MkldnnCPUQInt8.  We also extend the tensor methods to_mkldnn/to_dense to handle the conversion from QuantizedCPUQInt8 to MkldnnCPUQInt8, and linear_quantize/dequantize to handle the conversion between MkldnnCPUFloat and MkldnnCPUQInt8.  Methods q_scale/q_zero_point are supported. 

Following the QINT8 design, we also plan to add the MKL-DNN INT8 optimization pass for JIT,  and share the same calibration algorithm as Qtensor. We may contribute INT8 recipes for MKL-DNN Int8.

**Comment by Jianhui-Li** [comment]
https://github.com/pytorch/pytorch/pull/22390 implements this RFC.

**Comment by Jianhui-Li** [comment]
Update - With further discussion with Pytorch and FPGEMM team, we agreed to integrate MKL-DNN int8 with the same CPU FPGEMM integration interface, to avoid the complexity that Pytorch needs to integrate two libraries in two different ways.  Both use the same format NHWC and support same op semantics, so user doesn't worry about the tensor format conversion and op semantics difference.  The PR 26794 (https://github.com/pytorch/pytorch/pull/26794) attempted the new integration interface but it falls back to FBGEMM for the AQ inputs as MKL-DNN currently supports symmetric quantization (SQ).    DNNL v1.2 targets to have matmul primitive to support dynamic shape and asymmetric quantization with uint8 activation input.  We are looking into AQ support for conv down the road and expect to have a new PR with AQ support matched.  @jerryzh168

**Comment by david-macleod** [comment]
What is the current status on this? Will it eventually become available as an alternative backend for quantization along with FBGEMM and qnnpack? Should we expect significant speedups over FBGEMM (given that already used `avx512_vnni` instructions)

**Comment by jgong5** [comment]
> What is the current status on this? Will it eventually become available as an alternative backend for quantization along with FBGEMM and qnnpack? Should we expect significant speedups over FBGEMM (given that already used avx512_vnni instructions)

@david-macleod Yes. This is something we are working on right now. I won't say oneDNN (MKL-DNN) would be *significantly* faster than FBGEMM with avx512_vnni but it does show better end-to-end performance on some popular CNN models we tested. We will share more perf details in the PR. Moreover, you will get additional speedups from oneDNN with the most recent Intel HW features, e.g. AMX instructions on Sapphire Rapids CPU.

**Comment by Xia-Weiwen** [comment]
Update
**PR plan for MKLDNN int8 support**
- [ ] Add MKLDNN to qengine list
- [ ] Implement qconv
- [ ] Implement qlinear

(to be continued.)

**Comment by louie-tsai** [comment]
@aice-support
